### PR TITLE
cross entropy metrics and objective

### PR DESCRIPTION
**Pull request by olofer**
Extension to the objective and metric options
(sort-of merged classification on outcomes/regression on probabilities)

- standard cross entropy, evaluates on continuous labels in [0, 1]
- standard cross entropy with offset term (Kullback-Leibler divergence)
- special-weight cross-entropy that can be used to learn intensity functions (weights are lengths/duration)

### CODE REVIEW

File: src/metric/xentropy_metric.hpp [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    label_ = metadata.label();
+    weights_ = metadata.weights();
+
+    // ensure that labels are in interval [0, 1], interval ends included
+    if (label_ == nullptr) {
+       Log::Fatal("[%s]: label nullptr", __PRETTY_FUNCTION__);
+    }
+    for (data_size_t i = 0; i < num_data; ++i) {
+      if (label_[i] < 0.0f || label_[i] > 1.0f) {
+        Log::Fatal("[%s]: metric does not tolerate label [#%i] outside [0, 1]", __PRETTY_FUNCTION__, i);
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: what is ```__PRETTY_FUNCTION__``` ? it misses the definition

Comment (chivee) [inline src/metric/xentropy_metric.hpp]: `__PRETTY_FUNCTION__` was only defined by GNU/GCC. please note that MSVC didn't support this.

File: src/metric/xentropy_metric.hpp [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+  /*! \brief Sum of weights */
+  double sum_weights_;
+  /*! \brief Name of this metric */
+  std::vector<std::string> name_;
+};
+
+//
+// CrossEntropyMetric1 : "xentropy1" : (optional) weights have a different meaning than for "xentropy"
+//
+class CrossEntropyMetric1 : public Metric {
```
Comment (chivee) [inline src/metric/xentropy_metric.hpp]: Please using an self-explained class name for this `"intensity-weighted" ` metric

Comment (olofer) [inline src/metric/xentropy_metric.hpp]: Yes the class name can be improved.

File: src/metric/xentropy_metric.hpp [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+      Log::Fatal("[%s]: sum-of-weights = %f is non-positive", __func__, sum_weights_);
+    } else {
+      Log::Info("[%s:%s]: sum-of-weights = %f",  GetName()[0].c_str(), __func__, sum_weights_);
+    }
+  }
+
+  std::vector<double> Eval(const double* score, const ObjectiveFunction* objective) const override {
+    double sum_loss = 0.0f;
+    if (objective == nullptr) {
+      Log::Warning("[%s : %d]: objective nullptr", __FILE__, __LINE__);
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: I think this warning can be removed, since it is possible in python/R package.

File: src/metric/xentropy_metric.hpp [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+          objective->ConvertOutput(&score[i], &p);
+          sum_loss += XentLoss(label_[i], p) * weights_[i];
+        }
+      }
+    }
+    double loss = sum_loss / sum_weights_;
+    return std::vector<double>(1, loss);
+  }
+
+	const std::vector<std::string>& GetName() const override {
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: fix indent here

File: src/metric/xentropy_metric.hpp [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    double sum_loss = 0.0f;
+    if (objective == nullptr) {
+      Log::Warning("[%s : %d]: objective nullptr", __FILE__, __LINE__);
+      if (weights_ == nullptr) {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          double hhat = std::log(1.0f + std::exp(score[i])); // auto-convert
+          sum_loss += XentLoss1(label_[i], 1.0f, hhat);
+          //double p = 1.0f / (1.0f + std::exp(-score[i])); // auto-convert
+          //sum_loss += XentLoss(label_[i], p);
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: remove the useless code ?

File: src/metric/xentropy_metric.hpp [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+      }
+    } else {
+      // safe to not do anything
+    }
+  }
+
+  std::vector<double> Eval(const double* score, const ObjectiveFunction* objective) const override {
+    double sum_loss = 0.0f;
+    if (objective == nullptr) {
+      Log::Warning("[%s : %d]: objective nullptr", __FILE__, __LINE__);
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: remove the warning .

File: src/metric/xentropy_metric.hpp [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+        Log::Warning("[%s:%s]: minimum weight = %f is negative", GetName()[0].c_str(), __func__, min_weight);
+      }
+    }
+
+    // check weight sum
+    if (sum_weights_ <= 0.0f) {
+      Log::Fatal("[%s]: sum-of-weights = %f is non-positive", __func__, sum_weights_);
+    } else {
+      Log::Info("[%s:%s]: sum-of-weights = %f",  GetName()[0].c_str(), __func__, sum_weights_);
+    }
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: There are many duplicated code for the checking the ```label``` and ```weight``` in this file. To avoid that, I think you can define some inline static functions. like ```CheckLabel(const float* label, data_size_t num_data)``` .

File: src/objective/xentropy_objective.hpp [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        }
+        if (weights_[i] < min_weight_) min_weight_ = weights_[i];
+        if (weights_[i] > max_weight_) max_weight_ = weights_[i];
+      }
+      // Issue an info statement about this ratio; and possibly warn if excessively large ?!
+      double weight_ratio = max_weight_ / min_weight_;
+      Log::Info("[%s:%s]: min, max weights = %f, %f; ratio = %f", 
+                GetName(), __func__,
+                min_weight_, max_weight_,
+                weight_ratio);
```
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: the same duplicated code like in ```metric``` . Maybe these checks can be define in ```LightGBM/utils/Common.h```, and all metrics and objective can call them.

File: src/objective/xentropy_objective.hpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+    output[0] = 1.0f / (1.0f + std::exp(-input[0]));
+  }
+
+  std::string ToString() const override {
+    std::stringstream str_buf;
+    str_buf << GetName();
+    return str_buf.str();
+  }
+
+  bool BoostFromAverage() const override { return true; }
```
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: BoostFromAverage usually be used in regression.

File: src/objective/xentropy_objective.hpp [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  std::string ToString() const override {
+    std::stringstream str_buf;
+    str_buf << GetName();
+    return str_buf.str();
+  }
+
+  // TODO: default boost from different (weighted) average ?; not quite the same ...
+  bool BoostFromAverage() const override { return true; }
```
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: BoostFromAverage usually be used in regression.

File: src/objective/xentropy_objective.hpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+/*
+ * Implements gradients and hessians for the following point losses.
+ * Target y is anything in interval [0, 1].
+ *
+ * (1) CrossEntropy; "xentropy";
+ * 
+ * loss(y, p, w) = { -(1-y)*log(1-p)-y*log(p) }*w,
+ * with probability p = 1/(1+exp(-f)), where f is being boosted
+ *
+ * ConvertToOutput: f -> p
```
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: I think the ```binary``` objective already use the CrossEntropy

File: src/metric/xentropy_metric.hpp [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+    if (p > 0) hp += p * std::log(p);
+    double q = 1.0f - p;
+    if (q > 0) hp += q * std::log(q);
+    return hp;
+  }
+
+//
+// CrossEntropyMetric : "xentropy" : (optional) weights are used linearly
+//
+class CrossEntropyMetric : public Metric {
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: What is different between this and ```binary_logloss``` ?

File: src/metric/xentropy_metric.hpp [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+    if (1.0f - prob > log_arg_epsilon) {
+      b *= std::log(1.0f - prob);
+    } else {
+      b *= std::log(log_arg_epsilon);
+    }
+    return - (a + b);
+  }
+
+  // hhat >(=) 0 assumed; and weight > 0 required; but not checked here
+  inline static double XentLoss1(float label, float weight, double hhat) {
```
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: better naming ?

File: src/metric/xentropy_metric.hpp (hunk 12 above)
Comment (olofer) [inline src/metric/xentropy_metric.hpp]: Binary logloss is the special case of this when all targets are either 0 or 1.
This does "regression" on probabilities in general with continuous targets in [0, 1].
It can be thought of as "infinite-data" binary logloss when you have actual probabilities to give the learning algorithm. Binary logloss tries to count the number of pos. and neg. classes, whereas this never does because it makes no sense if you use the [0, 1] targets. The objective instead becomes a "distance" to a target distribution. Sometimes this function-fitting mode is what you want in my experience.

File: src/metric/xentropy_metric.hpp (hunk 7 above)
Comment (olofer) [inline src/metric/xentropy_metric.hpp]: The test/requirements are not exactly the same, but this can be improved.

File: src/metric/xentropy_metric.hpp (hunk 3 above)
Comment (olofer) [inline src/metric/xentropy_metric.hpp]: What exactly is the situation where this is called without an objective function?
It seems there is not enough metadata (in general) to treat a score correctly for a metric unless you know how the objective function score is parameterized; easy to misapply metrics yes?

File: src/objective/xentropy_objective.hpp (hunk 9 above)
Comment (olofer) [inline src/objective/xentropy_objective.hpp]: Seems correct here?

File: src/metric/xentropy_metric.hpp (hunk 13 above)
Comment (olofer) [inline src/metric/xentropy_metric.hpp]: Sure.

File: src/objective/xentropy_objective.hpp (hunk 11 above)
Comment (olofer) [inline src/objective/xentropy_objective.hpp]: Yes it should; it is a special case, see comment below.

File: src/objective/xentropy_objective.hpp (hunk 10 above)
Comment (olofer) [inline src/objective/xentropy_objective.hpp]: I think that's what it should be..

File: src/objective/xentropy_objective.hpp (hunk 9 above)
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: ok, got it

File: src/metric/xentropy_metric.hpp (hunk 7 above)
Comment (guolinke) [inline src/metric/xentropy_metric.hpp]: How about the checking for weights ?

File: src/objective/xentropy_objective.hpp (hunk 11 above)
Comment (guolinke) [inline src/objective/xentropy_objective.hpp]: okay, I see.

File: src/metric/xentropy_metric.hpp (hunk 7 above)
Comment (olofer) [inline src/metric/xentropy_metric.hpp]: Yeah I should try to arrange some commons for the weight checking..

**Comment by guolinke** [comment]
@olofer Thanks !
